File: packages/python/catalogmx/catalogs/mexico/salarios_minimos.py
"""
Mexican Minimum Wage Catalog

This module provides access to historical minimum wage data for Mexico.
"""
import json
from datetime import date
from pathlib import Path
from typing import List, Dict, Optional


class SalariosMinimos:
    """
    Catalog of Mexican minimum wages

    Provides historical minimum wage data from 2010 to present,
    with separate values for border zone and rest of country.
    """

    _data: Optional[List[Dict]] = None

    @classmethod
    def _load_data(cls) -> None:
        """Load minimum wage data from JSON file"""
        if cls._data is None:
            # Path: catalogmx/packages/python/catalogmx/catalogs/mexico/salarios_minimos.py
            # Target: catalogmx/packages/shared-data/mexico/salarios_minimos.json
            current_file = Path(__file__)
            shared_data_path = current_file.parent.parent.parent.parent.parent / 'shared-data' / 'mexico' / 'salarios_minimos.json'

            with open(shared_data_path, 'r', encoding='utf-8') as f:
                cls._data = json.load(f)

    @classmethod
    def get_data(cls) -> List[Dict]:
        """
        Get all minimum wage data

        :return: List of all minimum wage records
        """
        cls._load_data()
        return cls._data.copy()

    @classmethod
    def get_por_anio(cls, anio: int) -> Optional[Dict]:
        """
        Get minimum wage for a specific year

        :param anio: Year (e.g., 2024)
        :return: Minimum wage record or None if not found
        """
        cls._load_data()

        for record in cls._data:
            if record['año'] == anio:
                return record.copy()

        return None

    @classmethod
    def get_actual(cls) -> Optional[Dict]:
        """
        Get current minimum wage (latest year in data)

        :return: Latest minimum wage record
        """
        cls._load_data()

        if not cls._data:
            return None

        # Data is sorted by year descending, so first element is most recent
        return cls._data[0].copy()

    @classmethod
    def calcular_mensual(cls, diario: float, dias: int = 30) -> float:
        """
        Calculate monthly wage from daily wage

        :param diario: Daily wage amount
        :param dias: Number of days (default 30)
        :return: Monthly wage
        """
        return diario * dias

    @classmethod
    def calcular_anual(cls, diario: float, dias: int = 365) -> float:
        """
        Calculate annual wage from daily wage

        :param diario: Daily wage amount
        :param dias: Number of days (default 365)
        :return: Annual wage
        """
        return diario * dias

    @classmethod
    def get_por_zona(cls, anio: int, zona_frontera: bool = False) -> Optional[float]:
        """
        Get minimum wage for specific zone and year

        :param anio: Year
        :param zona_frontera: True for border zone, False for rest of country
        :return: Daily wage amount or None if not found
        """
        record = cls.get_por_anio(anio)
        if not record:
            return None

        if zona_frontera:
            return record.get('zona_frontera_norte')
        else:
            return record.get('resto_pais')

    @classmethod
    def get_incremento(cls, anio: int) -> Optional[float]:
        """
        Get percentage increment for a specific year

        :param anio: Year
        :return: Percentage increment or None if not found
        """
        record = cls.get_por_anio(anio)
        if not record:
            return None

        return record.get('incremento_porcentual')


# Convenience functions
def get_salario_actual() -> Optional[Dict]:
    """Get current minimum wage"""
    return SalariosMinimos.get_actual()


def get_salario_por_anio(anio: int) -> Optional[Dict]:
    """Get minimum wage for a specific year"""
    return SalariosMinimos.get_por_anio(anio)


def calcular_mensual(diario: float, dias: int = 30) -> float:
    """Calculate monthly wage from daily wage"""
    return SalariosMinimos.calcular_mensual(diario, dias)


# Export commonly used functions and classes
__all__ = [
    'SalariosMinimos',
    'get_salario_actual',
    'get_salario_por_anio',
    'calcular_mensual',
]

File: packages/typescript/src/catalogs/mexico/salarios-minimos.ts
import { loadCatalogArray } from '../../utils/catalog-loader';
import { SalarioMinimo } from '../../types';

/**
 * Mexican Minimum Wages Catalog
 * Historical data of minimum wages in Mexico from 2010 to present
 */
export class SalariosMinimos {
  private static _data: SalarioMinimo[] | null = null;

  private static loadData(): void {
    if (this._data !== null) return;
    this._data = loadCatalogArray<SalarioMinimo>('mexico/salarios_minimos.json');
  }

  static getData(): SalarioMinimo[] {
    this.loadData();
    return this._data!;
  }

  /**
   * Get minimum wage for a specific year
   */
  static getPorAño(año: number): SalarioMinimo | undefined {
    return this.getData().find((s) => s.año === año);
  }

  /**
   * Get current minimum wage (most recent year)
   */
  static getActual(): SalarioMinimo {
    const sorted = [...this.getData()].sort((a, b) => b.año - a.año);
    return sorted[0];
  }

  /**
   * Get minimum wage for a specific date
   */
  static getPorFecha(fecha: string | Date): SalarioMinimo | undefined {
    const date = typeof fecha === 'string' ? new Date(fecha) : fecha;
    const sorted = [...this.getData()].sort(
      (a, b) => new Date(b.vigencia_inicio).getTime() - new Date(a.vigencia_inicio).getTime()
    );

    return sorted.find((s) => new Date(s.vigencia_inicio) <= date);
  }

  /**
   * Get minimum wage value for a zone and year
   */
  static getValor(
    año: number,
    zona: 'frontera' | 'pais' | 'general' | 'a' | 'b' = 'pais'
  ): number | undefined {
    const salario = this.getPorAño(año);
    if (!salario) return undefined;

    switch (zona) {
      case 'frontera':
        return salario.zona_frontera_norte;
      case 'pais':
        return salario.resto_pais ?? salario.zona_general;
      case 'general':
        return salario.zona_general;
      case 'a':
        return salario.zona_a;
      case 'b':
        return salario.zona_b;
      default:
        return salario.resto_pais ?? salario.zona_general;
    }
  }

  static getUmaEquivalente(
    año: number,
    tipo: 'diario' | 'mensual' | 'anual' = 'diario'
  ): number | undefined {
    const salario = this.getPorAño(año);
    if (!salario) return undefined;

    switch (tipo) {
      case 'diario':
        return salario.uma_equivalente_diario;
      case 'mensual':
        return salario.uma_equivalente_mensual;
      case 'anual':
        return salario.uma_equivalente_anual;
      default:
        return salario.uma_equivalente_diario;
    }
  }

  static getFuenteUmaEquivalente(año: number): SalarioMinimo['fuente_uma_equivalente'] {
    const salario = this.getPorAño(año);
    return salario?.fuente_uma_equivalente;
  }

  /**
   * Calculate monthly minimum wage (assuming 30.4 days)
   */
  static calcularMensual(año: number, zona: 'frontera' | 'pais' = 'pais'): number | undefined {
    const diario = this.getValor(año, zona);
    return diario ? diario * 30.4 : undefined;
  }

  /**
   * Calculate annual minimum wage (assuming 365 days)
   */
  static calcularAnual(año: number, zona: 'frontera' | 'pais' = 'pais'): number | undefined {
    const diario = this.getValor(año, zona);
    return diario ? diario * 365 : undefined;
  }

  /**
   * Get historical evolution of minimum wages
   */
  static getHistorico(añoInicio: number, añoFin: number): SalarioMinimo[] {
    return this.getData()
      .filter((s) => s.año >= añoInicio && s.año <= añoFin)
      .sort((a, b) => a.año - b.año);
  }

  /**
   * Calculate percentage increase between two years
   */
  static calcularIncremento(
    añoInicial: number,
    añoFinal: number,
    zona: 'frontera' | 'pais' = 'pais'
  ): number | undefined {
    const inicial = this.getValor(añoInicial, zona);
    const final = this.getValor(añoFinal, zona);

    if (!inicial || !final) return undefined;
    return ((final - inicial) / inicial) * 100;
  }
}
